### PR TITLE
Review

### PR DESCRIPTION
**Pull request by poneill**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# markov-analysis
```
Comment (poneill) [inline README.md:1]: The general quality of this documentation is really good.  Consider adding a usage section at the top, though.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+>    combine text from two or more books, the random text you generate will blend
+>    the vocabulary and phrases from the sources in interesting ways.
+
+### Further Exploration
+
+My solution is likely not the most efficient, and since writing it I have become more familiar with Python.
+I would like to develop intuitions for how to determine which algorithms or data structures suit which applications,
+and how to efficiently implement these in Python.
+
+Some options:
```
Comment (poneill) [inline README.md:28]: how did this turn out, btw?

File: markov-analysis.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    another solution might be to just construct a list of each occurrence 
+
+    simple idea:
+    don't use word frequencies, just a set of suffixes
+    so the dict is a mapping P -> S 
+    (p_1, p_2, ..., p_n) -> (s_1, s_2, ..., s_n)
+
+    after finishing:
+    not sure why i ever thought of using a set of suffixes instead of a list,
+    since that would even preserve relative frequencies!
```
Comment (poneill) [inline markov-analysis.py:59]: right.

But what would happen if you used the list approach on a very large corpus, compared to just storing the frequencies directly?

File: markov-analysis.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+def generate_text(prefix_length, text_length, title='emma.txt'):
+    '''
+    Generates random text using the specified number of prefixes.
+
+    prefix_length: int, the number of prefix words to use
+    text_length: int, the length of the random text
+    title: string, title of the source book
+
+    int, int -> str
```
Comment (poneill) [inline markov-analysis.py:90]: we should talk about `mypy` at some point.

File: markov-analysis.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    word_list = []
+    for title in titles:
+        # WARNING: skip_header only works for emma.txt, Project Gutenberg books
+        #          do NOT have a standard header format
+        word_list = word_list + process_file(title, skip_header=False)
+    prefixes = prefix_map(word_list, prefix_length)
+    return list_to_string(random_text(prefixes, prefix_length, text_length))
+
+
+def generate_text(prefix_length, text_length, title='emma.txt'):
```
Comment (poneill) [inline markov-analysis.py:82]: wait is this function used at all?

File: markov-analysis.py (hunk 4 above)
Comment (poneill) [reply markov-analysis.py:90]: In python 3 you can explicitly write:

```def generate_text(prefix_length: int, text_length: int, title: str='emma.txt') -> str:```

and mypy can statically check them for you.

File: markov-analysis.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    text_length: int, the length of the random text
+    title: string, title of the source book
+
+    int, int -> str
+    '''
+    word_list = process_file(title, skip_header=True)
+    prefixes = prefix_map(word_list, prefix_length)
+    return list_to_string(random_text(prefixes, prefix_length, text_length))
+
+def list_to_string(word_list):
```
Comment (poneill) [inline markov-analysis.py:96]: howabout: `' '.join(word_list) + '.'`

File: markov-analysis.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    word_list: list
+
+    list -> str
+    '''
+    s = ''
+    for word in word_list:
+        s = s + ('%s ' % word)
+    return (s.strip() + '.')
+
+def random_text(prefix_map, prefix_length, text_length):
```
Comment (poneill) [inline markov-analysis.py:110]: best to give functions imperative names, e.g. `generate_random_text`

File: markov-analysis.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+    # TODO: This looks messy, is there a better way than changing so many types?
+
+    prefix_map: dict of tuples to lists
+    prefix_length: int, the length (in words) of the sequence of prefixes to use
+    m: int, the length (in words) of the list of random words
+
+    dict, int, int -> list
+    '''
+    random_text = list(random.choice(list(prefix_map.keys())))
```
Comment (poneill) [inline markov-analysis.py:123]: giving local vars the same name as your function is to be avoided

File: markov-analysis.py (hunk 7 above)
Comment (poneill) [reply markov-analysis.py:110]: function signature is surprising to me.  Isn't `prefix_length` redundant, given that all keys in `prefix_map` have that length?

File: markov-analysis.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+    while (len(random_text) < text_length):
+        prefix = tuple(random_text[(len(random_text)-prefix_length):len(random_text)])
```
Comment (poneill) [inline markov-analysis.py:126]: this is way too much logic on one line in general.

In this case, you could also shorten this to: 

```
prefix = tuple(random_text[-prefix_length:])
```
which means, "get me the last `prefix_length` elements of `random_text`".

File: markov-analysis.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    Returns a list of words randomly selected by prefix using the given 
+    prefix map.
+
+    # TODO: This looks messy, is there a better way than changing so many types?
+
+    prefix_map: dict of tuples to lists
+    prefix_length: int, the length (in words) of the sequence of prefixes to use
+    m: int, the length (in words) of the list of random words
+
+    dict, int, int -> list
```
Comment (poneill) [inline markov-analysis.py:121]: why would you want to return a list of strings here, rather than a string?

File: markov-analysis.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    '''
+    random_text = list(random.choice(list(prefix_map.keys())))
+
+    while (len(random_text) < text_length):
+        prefix = tuple(random_text[(len(random_text)-prefix_length):len(random_text)])
+        random_text.append(random.choice(prefix_map[prefix]))
+
+    return random_text
```
Comment (poneill) [inline markov-analysis.py:129]: `random_text` isn't really text-- it's a list of strings.  Would prefer the conversion to str happen here.

File: markov-analysis.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+def prefix_map(word_list, n):
```
Comment (poneill) [inline markov-analysis.py:131]: better, `make_prefix_map` or similar

File: markov-analysis.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+    '''
+    Takes a list of words and returns a dict mapping sequences of n words (which
+    we call the prefixes) to a list of all next words (which we call the 
+    suffixes).
+
+    word_list: list
+    n: int, the number of prefixes to consider
+    
+    list, int -> dict
```
Comment (poneill) [inline markov-analysis.py:140]: you can also specify the types of dicts.  If your function always returns dicts of a certain type, e.g. `Dict[Tuple[str], Tuple[Str]]` or whatever, better to specify that.

File: markov-analysis.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    '''
+    ''' TODO: This doesn't do quite what I want but can't figure out why
```
Comment (poneill) [inline markov-analysis.py:142]: I would prefer to handle this problem as follows.  The task has two parts:

1) scan a text and extract word statistics from it in the form of a `prefix_map`.
2) generate new text from a `prefix_map`.

The behavior you're trying to specify, i.e. how to initiate (2), properly belongs in step (2).

File: markov-analysis.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+        prefix_map = {(''):{word_list[0]}} # First word follows nothing
+        if n > 1: # Map starting prefix sequences of length <n
```
Comment (poneill) [inline markov-analysis.py:144]: if you want to seed your random text with a certain string, probably better to pass that string explicitly into the text generation function.

File: markov-analysis.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+            for i in range(1, n):
+                prefix_seq = tuple(word_list[1:i])
+                prefix_map[prefix_seq] = {word_list[i]}
+    '''
+    prefix_map = {}
+    # Maps prefix n-tuples to suffix list
+    for prefix_start in range(0, len(word_list)-n):
```
Comment (poneill) [inline markov-analysis.py:151]: better to iterate over words explicitly.  see `funcy.partition` https://funcy.readthedocs.io/en/stable/seqs.html

File: markov-analysis.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+    returns: list of all words in the file in order of appearance
+    """
+    word_list = []
+    fp = open(filename)
+
+    if skip_header:
+        skip_gutenberg_header(fp)
+
+    for line in fp:
+        process_line(line, word_list)
```
Comment (poneill) [inline markov-analysis.py:176]: my personal preference here would be to rewrite in a style more like:

`word_list += process_line(line)` 

or something.

implicit mutation of arguments is sketchy.

File: markov-analysis.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    return word_list
+
+# WARNING: This is only known to work with emma since Project Gutenberg books 
+#          do NOT have a standard header format.
+def skip_gutenberg_header(fp):
+    """
+    (Taken from chapter 13 of ThinkPython)
+    Reads from fp until it finds the line that ends the header.
+
+    TODO: Project Gutenberg books do not have a standard header. Figure out how
```
Comment (poneill) [inline markov-analysis.py:187]: this is a pretty common problem, and is tricky to do in a perfectly principled way.  If it were *really important* to get this right, I might:

- Sample N titles at random, group their headers into common types.  Handle all observed cases.
- Use Good-Turing estimation to estimate the probability that there is an unobserved type.  (As a rough rule of thumb, this goes as 3/N.)
- Sample more titles and iterate until you're satisfied with the probability of missing a type.

File: markov-analysis.py [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+    TODO: Some Project Gutenberg books have additional material after the book.
+          Figure out how to skip that.
+
+    fp: open file object
+    """
+    for line in fp:
+        if line.startswith('*END*THE SMALL PRINT!'):
+            break
+
+def process_line(line, word_list):
```
Comment (poneill) [inline markov-analysis.py:198]: I know this is part of the exercise, but in practice you'd probably want to use a proper tokenizer here.

Comment (poneill) [reply markov-analysis.py:198]: there's a surprising number of edge cases.

File: README.md (hunk 1 above)
Comment (jamais-vu) [reply README.md]: :+1:

File: README.md (hunk 2 above)
Comment (jamais-vu) [reply README.md:28]: i have not yet done anything listed under "further exploration"

File: markov-analysis.py (hunk 5 above)
Comment (jamais-vu) [reply markov-analysis.py:82]: `generate_text` was for part 2 of the exercise; `mashup` effectively generalizes it to multiple texts

File: markov-analysis.py (hunk 4 above)
Comment (jamais-vu) [reply markov-analysis.py:90]: i think i did actually start using `mypy` slightly prior to this, but stopped for this project since i was having too much trouble defining types for functions like `prefix_map`, which returns a `dict[list[str]: list[str]]` (i'm not sure of the proper way to write that)

File: README.md (hunk 1 above)
Comment (jamais-vu) [reply README.md:1]: will do

File: markov-analysis.py (hunk 3 above)
Comment (jamais-vu) [reply markov-analysis.py:59]: that would take up a lot of space. it would be worth implementing `prefix_map` as a map of n-tuples to some collection of suffixes with corresponding frequencies; then, `random_text` would use these frequencies to make a weighted random choice of suffix.

File: markov-analysis.py (hunk 7 above)
Comment (jamais-vu) [reply markov-analysis.py:110]: `prefix_length` is necessary for line 126, where we obtain the latest n-tuple in the text:

```python
prefix = tuple(random_text[(len(random_text)-prefix_length):len(random_text)]
```

but you are correct that we could obtain this length from `prefix_map` itself

File: markov-analysis.py (hunk 10 above)
Comment (jamais-vu) [reply markov-analysis.py:121]: i don't remember my motivation for doing this. i think, since we don't need all the words as a single `string` until the end, i may have wanted to do one of two things:

- delay converting to `string`, in case we ended up processing the text more later

- break up how much any individual function did

but in terms of refactoring this code, returning a `string` would be clearer and simpler.

File: markov-analysis.py (hunk 4 above)
Comment (jamais-vu) [reply markov-analysis.py:90]: i see you clarify this for me here: https://github.com/poneill/markov-analysis/pull/1/files#r253278953

File: markov-analysis.py (hunk 6 above)
Comment (jamais-vu) [reply markov-analysis.py:96]: done

File: README.md (hunk 1 above)
Comment (jamais-vu) [reply README.md:1]: done

File: markov-analysis.py (hunk 8 above)
Comment (jamais-vu) [reply markov-analysis.py:123]: done (by changing function name to `generate_random_text`, as you suggested in https://github.com/poneill/markov-analysis/pull/1#discussion_r253278687 )

File: markov-analysis.py (hunk 9 above)
Comment (jamais-vu) [reply markov-analysis.py:126]: done

File: markov-analysis.py (hunk 10 above)
Comment (jamais-vu) [reply markov-analysis.py:121]: done (now returns `str`)

File: markov-analysis.py (hunk 5 above)
Comment (jamais-vu) [reply markov-analysis.py:82]: done; i've removed `generate_text` since it has been superseded by `mashup`. i may change the name of `mashup` to something more descriptive.

File: markov-analysis.py (hunk 12 above)
Comment (jamais-vu) [reply markov-analysis.py:131]: done

File: markov-analysis.py (hunk 17 above)
Comment (jamais-vu) [reply markov-analysis.py:176]: done. i've rewritten `process_file` and `process_line` to avoid mutation, and changed their names to `file_to_list` and `line_to_list` respectively

File: markov-analysis.py (hunk 16 above)
Comment (jamais-vu) [reply markov-analysis.py:151]: done (and funcy looks to be very useful!)
